### DISCUSSION TITLE
Would you like to try this `run Keycloak server with a preconfigured application with just one command` ?

### DISCUSSION THREAD

**Discussion by leoporoli** [keycloak/keycloak#20249] [Feedback]
Hey, community!

Let me know guys what you think about this:

You can run the Keycloak server with a preconfigured app for testing purposes by following these steps:
1. [Install Kurtosis](https://docs.kurtosis.com/install) 
1. Execute this command `kurtosis run github.com/kurtosis-tech/keycloak-package` 

My intention with this [Kurtosis package](https://docs.kurtosis.com/concepts-reference/packages/) is to accelerate the onboarding process for Keycloak by automating the [Keycloak getting-started-with-Docker guide](https://www.keycloak.org/getting-started/getting-started-docker) so new users can try the server and the connection with an app in just a couple of seconds.

And this can be the first iteration, if we think that this is worth it, it can grow and add more benefits, like:
- Ingrate this package with [Postgres package](https://github.com/kurtosis-tech/postgres-package) using [composability](https://docs.kurtosis.com/explanations/reusable-environment-definitions#what-does-a-reusable-solution-look-like)
- Include a configure an SSL certificate
- Define your own app and connect it easily with the Keycloak server 

If you are curious about this package works and how it can be modified take a look at [the repository](https://github.com/kurtosis-tech/keycloak-package)

**Comment by slominskir** [comment]
Instead of Kurtosis you could just use docker compose, which you probably already have if you have Docker installed.  My quick starts skip the install Kurtosis step and instead are simply `docker compose up`

**Comment by leoporoli** [comment]
hey @slominskir I'm curious about your quick start with Docker compose, I'll like to test it. Could you share the link?

**Reply by slominskir** [reply]
https://github.com/slominskir/keycloak_oracle

**Comment by leoporoli** [comment]
Hey @slominskir! I just had a look at your Docker Compose files and I agree: the workflow I'm describing could indeed be done with docker-compose alongside the several Bash scripts you were using to check for service readiness and for creating resources in the server.

Kurtosis enables you to do the same thing, but without the overhead of creating, understanding, and maintaining several Bash scripts (these scripts may be overwhelming to a new user to reason about). Additionally, Kurtosis gives you the ability to compose your system together, meaning you can import this Keycloak package into another package to hook it up to a database and certificate in a first-class way.

Again, the end goal of my project is to help users get started using Keycloak faster and also help users build with Keycloak for their own projects! 

Would you like to test out the package I wrote? I would love to hear your feedback. Thank you!

**Comment by slominskir** [comment]
Hi @leoporoli, I agree that the Keycloak [container guide](https://www.keycloak.org/server/containers) could be improved and that using `docker run` is unwieldy for all but the simplest of demo, test, and development use cases.  Docker Compose and Kurtosis are both abstractions on top of Docker and inevitably users will have to dive into the details of whichever abstraction they choose so Kurtosis isn't necessarily saving people from being overwhelmed - they'd just be choosing to create, understand, and maintain with Kurtosis vs Docker Compose and bash.   

They're probably both good choices.    It's not immediately obvious if Kurtosis is better in some way, but I am familiar with Docker Compose and it seems to work well and allows you to compose your system together as well ([extends](https://docs.docker.com/compose/extends/), [profiles](https://docs.docker.com/compose/profiles/)).  Docker Compose and Bash are both well established and ubiquitous.   

There are some steps that could be taken by the Keycloak team to make Docker support in general friendlier.   I noticed starting with Keycloak 21 that the [image](https://quay.io/repository/keycloak/keycloak) no longer contains any package manager or fetch tool such as wget or curl.  This makes it very unfriendly for developers and all but necessitates they extend the image with their own Dockerfile.   A separate developer image may be in order if the current one is intended to stay production hardened.  

A built-in [Docker health check](https://docs.docker.com/engine/reference/builder/#healthcheck) would be nice too.   As soon as you start to test against a real-world Keycloak installation you immediately feel pain trying to keep container start order under control via healthchecks or custom entrypoint scripts.  For example [a typical use case](https://github.com/slominskir/keycloak_cluster) probably requires more than one instance of Keycloak, a database, a load balancer / proxy, and possibly LDAP or some other external user storage.

**Reply by leoporoli** [reply]
Yes, I agree about both are good choices, something that Kurtosis has as an extra benefit is that you can run the same package in Docker or Kubernetes and it's something that allows you to move from your local to the cloud without extra effort, probably for the current state of the package this is not worth it, but it could be worth it if this package grow by adding more services that could run in different instances/pods
